src/IJulia.jl

```diff
@@ -49,11 +52,140 @@ const depfile = joinpath(dirname(@__FILE__), "..", "deps", "deps.jl")
+        new(idents,header,content,parent_header,metadata)
+    end
+end
+
+verbose::Bool = IJULIA_DEBUG
+
+@kwdef mutable struct Kernel
+    verbose::Bool = IJULIA_DEBUG
+    inited::Bool = false
+    current_module::Module = Main
+
+    # These fields are special and are mirrored to their corresponding global
+    # variables.
+    In::Dict{Int, String} = Dict{Int, String}()
+    Out::Dict{Int, Any} = Dict{Int, Any}()
+    ans::Any = nothing
+    n::Int = 0
+    capture_stdout::Bool = true
+    capture_stderr::Bool = !IJULIA_DEBUG
+
+    postexecute_hooks::Vector{Function} = Function[]
+    preexecute_hooks::Vector{Function} = Function[]
+    posterror_hooks::Vector{Function} = Function[]
+    shutdown::Function = exit
+
+    # the following constants need to be initialized in init().
+    publish::RefValue{Socket} = Ref{Socket}()
+    raw_input::RefValue{Socket} = Ref{Socket}()
+    requests::RefValue{Socket} = Ref{Socket}()
+    control::RefValue{Socket} = Ref{Socket}()
+    heartbeat::RefValue{Socket} = Ref{Socket}()
+    heartbeat_context::RefValue{Context} = Ref{Context}()
+    profile::Dict{String, Any} = Dict{String, Any}()
+    connection_file::Union{String, Nothing} = nothing
+    read_stdout::RefValue{Base.PipeEndpoint} = Ref{Base.PipeEndpoint}()
+    read_stderr::RefValue{Base.PipeEndpoint} = Ref{Base.PipeEndpoint}()
+    socket_locks = Dict{Socket, ReentrantLock}()
+    hmacstate::RefValue{MbedTLS.MD{true}} = Ref{MbedTLS.MD{true}}()
+
+    requests_task::RefValue{Task} = Ref{Task}()
+    watch_stdout_task::RefValue{Task} = Ref{Task}()
+    watch_stderr_task::RefValue{Task} = Ref{Task}()
+    watch_stdout_timer::RefValue{Timer} = Ref{Timer}()
+    watch_stderr_timer::RefValue{Timer} = Ref{Timer}()
+
+    # name=>iobuffer for each stream ("stdout","stderr") so they can be sent in flush
+    bufs::Dict{String, IOBuffer} = Dict{String, IOBuffer}()
+    # max output per code cell is 512 kb by default
+    max_output_per_request::RefValue{Int} = Ref(1 << 19)
+
+    # Variable so that display can be done in the correct Msg context
+    execute_msg::Msg = Msg(["julia"], Dict("username"=>"jlkernel", "session"=>uuid4()), Dict())
+    # Variable tracking the number of bytes written in the current execution request
+    stdio_bytes::RefValue{Int} = Ref(0)
+    # Use an array to accumulate "payloads" for the execute_reply message
+    execute_payloads::Vector{Dict} = Dict[]
+
+    heartbeat_threadid::Vector{Int} = zeros(Int, 128) # sizeof(uv_thread_t) <= 8 on Linux, OSX, Win
+
+    # queue of objects to display at end of cell execution
+    displayqueue::Vector{Any} = Any[]
+end
+
+function Base.setproperty!(value::Kernel, name::Symbol, x)
+    # These fields are isbits types and need to be assigned explicitly to their
+    # global counterparts.
+    if name ∈ (:ans, :n, :In, :Out, :inited, :verbose)
+        setproperty!(IJulia, name, x)
+    end
+
+    setfield!(value, name, x)
+end
+
+function Base.close(kernel::Kernel)
```

Comment:
should be called by a `finalizer`?

Reply:
Unfortunately that wouldn't be easy because it `wait`'s on the stdio tasks so it might yield. I think it's ok to rely on the do-constructor closing it in the tests.